spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,155 @@
+    Identifier backupIdentifier = Identifier.of(sourceTableName().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableName(), backupIdentifier);
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException("Cannot find table to migrate", e);
+    } catch (TableAlreadyExistsException e) {
+      throw new IllegalArgumentException("Cannot rename migration source to backup name", e);
+    }
+
+    StagedTable stagedTable = null;
+    boolean threw = true;
+    try {
+      Map<String, String> newTableProperties = new HashMap<>();
+      newTableProperties.put(TableCatalog.PROP_PROVIDER, "iceberg");
+      newTableProperties.putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava());
+      newTableProperties.putAll(tableLocationProperties(sourceTableLocation()));
+      newTableProperties.putAll(additionalProperties());
+
+      stagedTable = destCatalog().stageCreate(Identifier.of(
+          destTableName().namespace(),
+          destTableName().name()),
+          v1SourceTable().schema(),
+          sourcePartitionSpec(),
+          newTableProperties);
+
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!stagedTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        assignDefaultTableNameMapping(icebergTable);
+      }
+
+      String stagingLocation = stagedTable.properties().get(TableProperties.WRITE_METADATA_LOCATION);
+
+      LOG.info("Beginning migration of {} using metadata location {}", sourceTableName(), stagingLocation);
+
+      Some<String> backupNamespace = Some.apply(backupIdentifier.namespace()[0]);
+      TableIdentifier v1BackupIdentifier = new TableIdentifier(backupIdentifier.name(), backupNamespace);
+      SparkTableUtil.importSparkTable(spark(), v1BackupIdentifier, icebergTable, stagingLocation);
+
+      stagedTable.commitStagedChanges();
+      threw = false;
+
+    } catch (TableAlreadyExistsException tableAlreadyExistsException) {
+      throw new IllegalArgumentException("Cannot migrate because a table was created with the same name as the " +
+          "original was created after we renamed the original table to a backup identifier.",
+          tableAlreadyExistsException);
+    } catch (NoSuchNamespaceException noSuchNamespaceException) {
+      throw new IllegalArgumentException("Cannot migrate because the namespace for the table no longer exists after" +
+          "we renamed the original table to a backup identifier.",
+          noSuchNamespaceException);
+    } finally {
+
+      if (threw) {
+        LOG.error("Error when attempting perform migration changes, aborting table creation and restoring backup.");
+        try {
+          destCatalog().renameTable(backupIdentifier, sourceTableName());
+        } catch (NoSuchTableException nstException) {
+          throw new IllegalArgumentException("Cannot restore backup, the backup cannot be found", nstException);
```

Comment:
Do we really want to throw a new exception rather than rethrwoing the original exception?

I kind of liked our internal utility `tryWithSafeFinallyAndFailureCallbacks` that we copied from Spark for such things.

Reply:
I don't think so. I think this should log and allow the original exception to propagate. It would also be nice to add these as suppressed exceptions.

Reply:
@RussellSpitzer, our `tryWithSafeFinallyAndFailureCallbacks` does add suppressed exceptions. We can address this in a follow-up but we need to fix other major points.

Reply:
I think we should log the errors for now and update to add suppressed exceptions later.

So for now, I think the runtime exceptions here should be changed to log messages.